File: apps/ios/BananaTape/BananaTape/NativeImageComposer.swift
import Foundation
import UIKit

struct NativeImageMetadata: Equatable {
    let width: Int
    let height: Int
    let byteCount: Int
    let mimeType: String
}

struct NativeComposedImage: Equatable {
    let fileURL: URL
    let metadata: NativeImageMetadata
}

struct NativeExportPreview: Equatable {
    let source: NativeImageMetadata
    let annotated: NativeImageMetadata
    let mask: NativeImageMetadata
    let canvasSize: EditorSize
}

struct NativeImageCompositionResult: Equatable {
    let original: NativeImageMetadata
    let annotated: NativeComposedImage
    let mask: NativeComposedImage
    let exportPreview: NativeExportPreview
}

enum NativeImageCompositionError: Error, Equatable {
    case unreadableSource
    case imageTooLarge(maxPixels: Int, actualPixels: Int)
    case renderFailed

    var userMessage: String {
        switch self {
        case .unreadableSource, .renderFailed:
            return "This image could not be prepared for export."
        case .imageTooLarge:
            return "This image is too large to prepare on this device."
        }
    }
}

struct NativeImageCompositionRequest: Equatable {
    let sourceURL: URL
    let annotations: CanvasAnnotations
    let outputDirectory: URL
}

struct NativeImageComposer {
    let maxPixelCount: Int

    init(maxPixelCount: Int = 16_777_216) {
        self.maxPixelCount = maxPixelCount
    }

    func compose(_ request: NativeImageCompositionRequest) -> Result<NativeImageCompositionResult, NativeImageCompositionError> {
        guard let sourceData = try? Data(contentsOf: request.sourceURL),
              let image = UIImage(contentsOfFile: request.sourceURL.path) ?? UIImage(data: sourceData) else {
            return .failure(.unreadableSource)
        }
        let size = pixelSize(for: image)
        let pixels = Int(size.width * size.height)
        guard pixels <= maxPixelCount else {
            return .failure(.imageTooLarge(maxPixels: maxPixelCount, actualPixels: pixels))
        }
        guard let annotatedData = renderAnnotated(image: image, size: size, annotations: request.annotations), let maskData = renderMask(size: size, annotations: request.annotations) else {
            return .failure(.renderFailed)
        }
        return writeResult(request: request, sourceData: sourceData, size: size, annotatedData: annotatedData, maskData: maskData)
    }

    private func writeResult(request: NativeImageCompositionRequest, sourceData: Data, size: CGSize, annotatedData: Data, maskData: Data) -> Result<NativeImageCompositionResult, NativeImageCompositionError> {
        do {
            try FileManager.default.createDirectory(at: request.outputDirectory, withIntermediateDirectories: true)
            let annotatedURL = request.outputDirectory.appendingPathComponent("annotated.png")
            let maskURL = request.outputDirectory.appendingPathComponent("mask.png")
            try annotatedData.write(to: annotatedURL, options: .atomic)
            try maskData.write(to: maskURL, options: .atomic)
            let width = Int(size.width)
            let height = Int(size.height)
            let original = NativeImageMetadata(width: width, height: height, byteCount: sourceData.count, mimeType: mimeType(for: request.sourceURL))
            let annotated = NativeImageMetadata(width: width, height: height, byteCount: annotatedData.count, mimeType: "image/png")
            let mask = NativeImageMetadata(width: width, height: height, byteCount: maskData.count, mimeType: "image/png")
            return .success(NativeImageCompositionResult(original: original, annotated: NativeComposedImage(fileURL: annotatedURL, metadata: annotated), mask: NativeComposedImage(fileURL: maskURL, metadata: mask), exportPreview: NativeExportPreview(source: original, annotated: annotated, mask: mask, canvasSize: EditorSize(width: Double(width), height: Double(height)))))
        } catch {
            return .failure(.renderFailed)
        }
    }

    private func renderAnnotated(image: UIImage, size: CGSize, annotations: CanvasAnnotations) -> Data? {
        UIGraphicsImageRenderer(size: size, format: rendererFormat()).pngData { context in
            image.draw(in: CGRect(origin: .zero, size: size))
            drawAnnotations(annotations, size: size, context: context.cgContext, mode: .annotated)
        }
    }

    private func renderMask(size: CGSize, annotations: CanvasAnnotations) -> Data? {
        UIGraphicsImageRenderer(size: size, format: rendererFormat()).pngData { context in
            UIColor.white.setFill()
            context.fill(CGRect(origin: .zero, size: size))
            context.cgContext.setBlendMode(.clear)
            drawAnnotations(annotations, size: size, context: context.cgContext, mode: .mask)
            context.cgContext.setBlendMode(.normal)
        }
    }

    private enum RenderMode {
        case annotated
        case mask
    }

    private func drawAnnotations(_ annotations: CanvasAnnotations, size: CGSize, context: CGContext, mode: RenderMode) {
        annotations.paths.forEach { draw($0, size: size, context: context) }
        annotations.boxes.forEach { draw($0, size: size, context: context, mode: mode) }
        if mode == .annotated { annotations.memos.forEach { draw($0, size: size) } }
    }

    private func draw(_ path: DrawingPath, size: CGSize, context: CGContext) {
        guard let first = path.points.first else { return }
        let cgPath = CGMutablePath()
        cgPath.move(to: canvasPoint(first, size: size))
        path.points.dropFirst().forEach { cgPath.addLine(to: canvasPoint($0, size: size)) }
        context.setStrokeColor(UIColor(hex: path.color).cgColor)
        context.setLineWidth(max(path.strokeWidth, 1))
        context.setLineCap(.round)
        context.setLineJoin(.round)
        context.addPath(cgPath)
        context.strokePath()
        if path.tool == .arrow, let last = path.points.last, path.points.count > 1 {
            drawArrowHead(from: path.points[path.points.count - 2], to: last, size: size, color: UIColor(hex: path.color), context: context)
        }
    }

    private func draw(_ box: BoundingBox, size: CGSize, context: CGContext, mode: RenderMode) {
        let rect = CGRect(x: box.x * size.width, y: box.y * size.height, width: box.width * size.width, height: box.height * size.height)
        if mode == .mask {
            context.fill(rect)
        } else {
            context.setStrokeColor(UIColor(hex: box.color).cgColor)
            context.setLineWidth(3)
            context.stroke(rect)
        }
    }

    private func draw(_ memo: TextMemo, size: CGSize) {
        UIColor(hex: memo.color).setFill()
        UIBezierPath(roundedRect: CGRect(x: memo.x * size.width, y: memo.y * size.height, width: 96, height: 48), cornerRadius: 8).fill()
    }

    private func drawArrowHead(from start: EditorPoint, to end: EditorPoint, size: CGSize, color: UIColor, context: CGContext) {
        let startPoint = canvasPoint(start, size: size)
        let endPoint = canvasPoint(end, size: size)
        let angle = atan2(endPoint.y - startPoint.y, endPoint.x - startPoint.x)
        let length = CGFloat(14)
        let spread = CGFloat.pi / 7
        let head = CGMutablePath()
        head.move(to: endPoint)
        head.addLine(to: CGPoint(x: endPoint.x - length * cos(angle - spread), y: endPoint.y - length * sin(angle - spread)))
        head.move(to: endPoint)
        head.addLine(to: CGPoint(x: endPoint.x - length * cos(angle + spread), y: endPoint.y - length * sin(angle + spread)))
        context.setStrokeColor(color.cgColor)
        context.setLineWidth(3)
        context.addPath(head)
        context.strokePath()
    }

    private func pixelSize(for image: UIImage) -> CGSize {
        if let cgImage = image.cgImage {
            return CGSize(width: cgImage.width, height: cgImage.height)
        }
        return CGSize(width: image.size.width * image.scale, height: image.size.height * image.scale)
    }

    private func canvasPoint(_ point: EditorPoint, size: CGSize) -> CGPoint {
        CGPoint(x: point.x * size.width, y: point.y * size.height)
    }

    private func rendererFormat() -> UIGraphicsImageRendererFormat {
        let format = UIGraphicsImageRendererFormat.default()
        format.opaque = false
        format.scale = 1
        return format
    }

    private func mimeType(for url: URL) -> String {
        let ext = url.pathExtension.lowercased()
        return ext == "jpg" || ext == "jpeg" ? "image/jpeg" : "image/png"
    }
}

private extension UIColor {
    convenience init(hex: String) {
        let trimmed = hex.trimmingCharacters(in: CharacterSet(charactersIn: "#"))
        let value = UInt32(trimmed, radix: 16) ?? 0x0D99FF
        self.init(red: CGFloat((value >> 16) & 0xFF) / 255, green: CGFloat((value >> 8) & 0xFF) / 255, blue: CGFloat(value & 0xFF) / 255, alpha: 1)
    }
}

File: apps/android/app/src/main/java/app/bananatape/mobile/editor/NativeImageComposer.kt
package app.bananatape.mobile.editor

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path as AndroidPath
import android.graphics.RectF
import java.nio.file.Files
import java.nio.file.Path
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.sin

data class NativeImageMetadata(
    val width: Int,
    val height: Int,
    val byteCount: Int,
    val mimeType: String,
)

data class NativeComposedImage(
    val filePath: Path,
    val metadata: NativeImageMetadata,
)

data class NativeExportPreview(
    val source: NativeImageMetadata,
    val annotated: NativeImageMetadata,
    val mask: NativeImageMetadata,
    val canvasSize: EditorSize,
)

data class NativeImageCompositionResult(
    val original: NativeImageMetadata,
    val annotated: NativeComposedImage,
    val mask: NativeComposedImage,
    val exportPreview: NativeExportPreview,
)

sealed class NativeImageCompositionError(val userMessage: String) {
    data object UnreadableSource : NativeImageCompositionError("This image could not be prepared for export.")
    data object RenderFailed : NativeImageCompositionError("This image could not be prepared for export.")
    data class ImageTooLarge(val maxPixels: Int, val actualPixels: Int) : NativeImageCompositionError("This image is too large to prepare on this device.")
}

data class NativeImageCompositionRequest(
    val sourcePath: Path,
    val annotations: CanvasAnnotations,
    val outputDirectory: Path,
)

sealed class NativeImageCompositionOutcome {
    data class Success(val result: NativeImageCompositionResult) : NativeImageCompositionOutcome()
    data class Failure(val error: NativeImageCompositionError) : NativeImageCompositionOutcome()
}

interface NativeImageRenderer {
    fun metadata(sourcePath: Path): NativeImageMetadata?
    fun annotatedPng(sourcePath: Path, annotations: CanvasAnnotations): ByteArray?
    fun maskPng(size: EditorSize, annotations: CanvasAnnotations): ByteArray?
}

class AndroidBitmapImageRenderer : NativeImageRenderer {
    override fun metadata(sourcePath: Path): NativeImageMetadata? {
        val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeFile(sourcePath.toString(), options)
        if (options.outWidth <= 0 || options.outHeight <= 0) return null
        return NativeImageMetadata(options.outWidth, options.outHeight, Files.size(sourcePath).toInt(), mimeType(sourcePath))
    }

    override fun annotatedPng(sourcePath: Path, annotations: CanvasAnnotations): ByteArray? {
        val source = BitmapFactory.decodeFile(sourcePath.toString()) ?: return null
        val bitmap = source.copy(Bitmap.Config.ARGB_8888, true)
        val canvas = Canvas(bitmap)
        drawAnnotations(canvas, bitmap.width.toFloat(), bitmap.height.toFloat(), annotations, includeMemos = true, mask = false)
        return bitmap.toPngBytes()
    }

    override fun maskPng(size: EditorSize, annotations: CanvasAnnotations): ByteArray? {
        val bitmap = Bitmap.createBitmap(size.width.toInt(), size.height.toInt(), Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        canvas.drawColor(Color.WHITE)
        drawAnnotations(canvas, size.width.toFloat(), size.height.toFloat(), annotations, includeMemos = false, mask = true)
        return bitmap.toPngBytes()
    }

    private fun drawAnnotations(canvas: Canvas, width: Float, height: Float, annotations: CanvasAnnotations, includeMemos: Boolean, mask: Boolean) {
        annotations.paths.forEach { drawPath(canvas, width, height, it, mask) }
        annotations.boxes.forEach { drawBox(canvas, width, height, it, mask) }
        if (includeMemos) annotations.memos.forEach { drawMemo(canvas, width, height, it) }
    }

    private fun drawPath(canvas: Canvas, width: Float, height: Float, path: DrawingPath, mask: Boolean) {
        val first = path.points.firstOrNull() ?: return
        val androidPath = AndroidPath().apply {
            moveTo((first.x * width).toFloat(), (first.y * height).toFloat())
            path.points.drop(1).forEach { lineTo((it.x * width).toFloat(), (it.y * height).toFloat()) }
        }
        val paint = strokePaint(if (mask) Color.TRANSPARENT else parseColor(path.color), max(path.strokeWidth.toFloat(), 1f))
        if (mask) paint.xfermode = android.graphics.PorterDuffXfermode(android.graphics.PorterDuff.Mode.CLEAR)
        canvas.drawPath(androidPath, paint)
        if (path.tool == DrawingTool.ARROW && path.points.size > 1) drawArrowHead(canvas, width, height, path, paint)
    }

    private fun drawBox(canvas: Canvas, width: Float, height: Float, box: BoundingBox, mask: Boolean) {
        val rect = RectF((box.x * width).toFloat(), (box.y * height).toFloat(), ((box.x + box.width) * width).toFloat(), ((box.y + box.height) * height).toFloat())
        val paint = if (mask) fillPaint(Color.TRANSPARENT) else strokePaint(parseColor(box.color), 3f)
        if (mask) paint.xfermode = android.graphics.PorterDuffXfermode(android.graphics.PorterDuff.Mode.CLEAR)
        canvas.drawRect(rect, paint)
    }

    private fun drawMemo(canvas: Canvas, width: Float, height: Float, memo: TextMemo) {
        canvas.drawRoundRect(RectF((memo.x * width).toFloat(), (memo.y * height).toFloat(), (memo.x * width).toFloat() + 96f, (memo.y * height).toFloat() + 48f), 8f, 8f, fillPaint(parseColor(memo.color)))
    }

    private fun drawArrowHead(canvas: Canvas, width: Float, height: Float, path: DrawingPath, paint: Paint) {
        val start = path.points[path.points.size - 2]
        val end = path.points.last()
        val startX = start.x * width
        val startY = start.y * height
        val endX = end.x * width
        val endY = end.y * height
        val angle = atan2(endY - startY, endX - startX)
        val length = 14.0
        val spread = Math.PI / 7.0
        canvas.drawLine(endX.toFloat(), endY.toFloat(), (endX - length * cos(angle - spread)).toFloat(), (endY - length * sin(angle - spread)).toFloat(), paint)
        canvas.drawLine(endX.toFloat(), endY.toFloat(), (endX - length * cos(angle + spread)).toFloat(), (endY - length * sin(angle + spread)).toFloat(), paint)
    }

    private fun strokePaint(color: Int, strokeWidth: Float): Paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        this.color = color
        style = Paint.Style.STROKE
        this.strokeWidth = strokeWidth
        strokeCap = Paint.Cap.ROUND
        strokeJoin = Paint.Join.ROUND
    }

    private fun fillPaint(color: Int): Paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        this.color = color
        style = Paint.Style.FILL
    }

    private fun Bitmap.toPngBytes(): ByteArray {
        val output = java.io.ByteArrayOutputStream()
        compress(Bitmap.CompressFormat.PNG, 100, output)
        return output.toByteArray()
    }

    private fun parseColor(value: String): Int = runCatching { Color.parseColor(value) }.getOrDefault(Color.rgb(13, 153, 255))

    private fun mimeType(path: Path): String {
        val name = path.fileName.toString().lowercase()
        return if (name.endsWith(".jpg") || name.endsWith(".jpeg")) "image/jpeg" else "image/png"
    }
}

class NativeImageComposer(
    private val renderer: NativeImageRenderer,
    private val maxPixelCount: Int = 16_777_216,
) {
    fun compose(request: NativeImageCompositionRequest): NativeImageCompositionOutcome {
        val source = renderer.metadata(request.sourcePath) ?: return NativeImageCompositionOutcome.Failure(NativeImageCompositionError.UnreadableSource)
        val pixels = source.width * source.height
        if (pixels > maxPixelCount) return NativeImageCompositionOutcome.Failure(NativeImageCompositionError.ImageTooLarge(maxPixelCount, pixels))
        val annotatedData = renderer.annotatedPng(request.sourcePath, request.annotations) ?: return NativeImageCompositionOutcome.Failure(NativeImageCompositionError.RenderFailed)
        val maskData = renderer.maskPng(EditorSize(source.width.toDouble(), source.height.toDouble()), request.annotations) ?: return NativeImageCompositionOutcome.Failure(NativeImageCompositionError.RenderFailed)
        return runCatching {
            Files.createDirectories(request.outputDirectory)
            val annotatedPath = request.outputDirectory.resolve("annotated.png")
            val maskPath = request.outputDirectory.resolve("mask.png")
            Files.write(annotatedPath, annotatedData)
            Files.write(maskPath, maskData)
            val annotated = NativeImageMetadata(source.width, source.height, annotatedData.size, "image/png")
            val mask = NativeImageMetadata(source.width, source.height, maskData.size, "image/png")
            NativeImageCompositionOutcome.Success(
                NativeImageCompositionResult(
                    original = source,
                    annotated = NativeComposedImage(annotatedPath, annotated),
                    mask = NativeComposedImage(maskPath, mask),
                    exportPreview = NativeExportPreview(source, annotated, mask, EditorSize(source.width.toDouble(), source.height.toDouble())),
                ),
            )
        }.getOrElse { NativeImageCompositionOutcome.Failure(NativeImageCompositionError.RenderFailed) }
    }
}
